Client/Frontend/Browser/BrowserViewController.swift

```diff
@@ -766,40 +768,88 @@ extension BrowserViewController: SearchViewControllerDelegate {
     }
 }
 
-extension BrowserViewController: UIScrollViewDelegate {
+class ScrollRecognizer : UIGestureRecognizer {
+    weak var bvc: BrowserViewController?
+    private var previousScroll: CGPoint?
 
-    func scrollViewWillBeginDragging(scrollView: UIScrollView) {
-        self.previousScroll = scrollView.contentOffset
+    override func canPreventGestureRecognizer(preventedGestureRecognizer: UIGestureRecognizer!) -> Bool {
+        return false
     }
 
-    // Careful! This method can be called multiple times concurrently.
-    func scrollViewDidScroll(scrollView: UIScrollView) {
-        if let tab = tabManager.selectedTab, let prev = self.previousScroll, let startOffset = self.startOffset {
-            let dy = prev.y - scrollView.contentOffset.y
-
-            if !tab.loading &&
-                scrollView.contentOffset.y > -startOffset  &&
-                scrollView.contentOffset.y < scrollView.contentSize.height {
-                self.scrollFooter(dy)
-                self.scrollHeader(dy)
-                self.scrollReader(dy)
+    override func canBePreventedByGestureRecognizer(preventingGestureRecognizer: UIGestureRecognizer!) -> Bool {
+        if preventingGestureRecognizer.description.rangeOfString("UIWebTouchEventsGestureRecognizer") != nil {
+            // We normally wouldn't want to allow touch events a chance to prevent this, but doing so can result.
+            // in lock ups of the entire UI.
+            if preventingGestureRecognizer.state == .Ended {
+                return true
             }
+            return false;
+        }
+        return preventingGestureRecognizer.description.rangeOfString("UIScrollViewPanGestureRecognizer") == nil
+    }
+
+    override init(target: AnyObject, action: Selector) {
+        super.init(target: target, action: action)
+    }
 
-            self.previousScroll = scrollView.contentOffset
+    override func touchesBegan(touches: Set<NSObject>!, withEvent event: UIEvent!) {
+        if touches.count == 1 {
+            if let touch = touches.first as? UITouch {
+                previousScroll = touch.locationInView(bvc?.view)
+            }
+        } else {
+            // If a second finer comes down, we'll stop dragging
+            endDragging()
+        }
+        super.touchesBegan(touches, withEvent: event)
+    }
+
+    override func touchesMoved(touches: Set<NSObject>!, withEvent event: UIEvent!) {
+        if previousScroll != nil {
+            if let touch = touches.first as? UITouch,
+                let tab = bvc?.tabManager.selectedTab,
+                let prev = self.previousScroll,
+                let startOffset = self.bvc?.startOffset {
+                    let offset = touch.locationInView(bvc?.view)
+                    let dy = offset.y - prev.y
+
+                    if !tab.loading &&
+                        offset.y > -startOffset  &&
+                        offset.y < bvc!.tabManager.selectedTab!.webView.scrollView.contentSize.height {
```

Comment:
I actually have a fix for this statement on a branch I'm working on where if the content height is near the size of the frame, scrolling goes all wonky because technically we only want to scroll if offset.y < bvc!.tabManager.selectedTab!.webView.scrollView.contentSize.height - scrollView.frame.size.height
Here's the bug: https://bugzilla.mozilla.org/show_bug.cgi?id=1154927

Also I we can clean this up since we already unwrapped the selected tab so you should be able to use tab directly instead of bvc!.tabManager.selected!